Feature: Dismiss all notifications button

aliaksandrb opened the pull request:
Hi!

Have added the "Dismiss All" button at user dashboard.
When user clicks on it and confirm the action,
all active (dimissed: false) notices for current user
would be automaticly dismissed.

Notices table becomes empty.

This button does not shown if there are no notices.

Feature: #9726
Link: https://labs.riseup.net/code/issues/9726

![dismis_all_message_workflow](https://cloud.githubusercontent.com/assets/526022/8633813/28a16f4a-27e6-11e5-983c-04325e2b20f6.png)

---
Any notices or recommendations?

azul commented:
Wow! I've been thinking about looking into this as i felt it would be useful. And out of the sudden :sparkles: there's a pull request for it :clap: :smile:

aliaksandrb commented:
It is all the opensource magic! :)

azul commented on app/views/me/notices/index.html.haml:
```diff
@@ -1,12 +1,24 @@
+%div
+  %span.h2.first=:notices.t
+  - if @notices.total_pages < 2 && @notices.any?
+    %span.h2.dissmiss-all-notices-button.dismiss-all-wrapper.pull-right
```
Is there a reason to have the `dissmiss-all-notices-button` class here? If there is one - why is it not applied below?

(just asking out of curiosity :thought_balloon: )

aliaksandrb replied:
It is special class to handle inline right-aligned position for the button in case when there is not pagination block.
Without it button displays near the `Notices` title or crosses the notices list table.

azul replied:
hmm... i searched the diff page on github for "dissmiss-all-notices-button" and i only found this one occurance. Am i missing something?

aliaksandrb replied:
Oh, my fault, have committed into git ignored file! :D

Let me fix this and other typo issues a bit later.
On Jul 11, 2015 5:35 PM, "azul" <notifications@github.com> wrote:

> In app/views/me/notices/index.html.haml
> <https://github.com/riseuplabs/crabgrass-core/pull/269#discussion_r34413164>
> :
>
> > @@ -1,12 +1,24 @@
> > -.h2.first=:notices.t
> > +%div
> > +  %span.h2.first=:notices.t
> > +  - if @notices.total_pages < 2 && @notices.any?
> > +    %span.h2.dissmiss-all-notices-button.dismiss-all-wrapper.pull-right
>
> hmm... i searched the diff page on github for
> "dissmiss-all-notices-button" and i only found this one occurance. Am i
> missing something?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/riseuplabs/crabgrass-core/pull/269/files#r34413164>.
>

elijh replied:
you probably know this, but you can update this same PR by doing a `push --force`. e.g.

    git checkout hide_all_notifications_link
    git push --force aliaksandrb hide_all_notifications_link

where `aliaksandrb` is the name of the remote you have added for your github account.

aliaksandrb replied:
Yeah, of course. Use it every time :)

azul commented on app/views/me/notices/index.html.haml:
```diff
@@ -1,12 +1,24 @@
+  %div
+    .pagination-wrapper.pull-left
+      = pagination_links @notices
+    - if @notices.total_pages > 1
```
Could we avoid having two different cases that add the button?
Wouldn't this place here work even if there are no pagination links?

aliaksandrb replied:
It depends on the place where we would like to place a new button / link maybe.
Current implementation place it at right corner of table, and the problem with it, that `Notices` title and pagination block have different display options and size.
So button placed differently if we use one css class for both cases..

By the way we could place it near the `Notices` title, for both cases, then it could be simplified, but it looks a bit overloaded for me then.

What about pagination: I believe that pagination block does not shown for pages less than 2, because it has no sense then.

azul replied:
How about removing the upper pagination block? I don't see why we would need it. You scroll down to read all notices and then move on to the next page.

aliaksandrb replied:
It is a nice point! I was wondering why everywhere in crabgrass there are
double pagination block, which is uncommon in those days.
I even through that it is a bug, but after found a commit where it was made
intentionally, so decided to leave it as it is :)

It appears in many places, like private messages for example.

Personally don't like duplicated pagination too :)
On Jul 11, 2015 5:34 PM, "azul" <notifications@github.com> wrote:

> In app/views/me/notices/index.html.haml
> <https://github.com/riseuplabs/crabgrass-core/pull/269#discussion_r34413148>
> :
>
> >  - if @notices.empty?
> >    %ul.shaded
> >      %li= :none.t
> >  - else
> > -  = pagination_links @notices
> > +  %div
> > +    .pagination-wrapper.pull-left
> > +      = pagination_links @notices
> > +    - if @notices.total_pages > 1
>
> How about removing the upper pagination block? I don't see why we would
> need it. You scroll down to read all notices and then move on to the next
> page.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/riseuplabs/crabgrass-core/pull/269/files#r34413148>.
>

azul replied:
For me it depends. For example for the comments below a page it kind of makes sense to me. The space there is not as 'valuable' anyway and the list of comments is long. But here i think it's better to save the space and have less logic in the view.

azul commented on app/helpers/me/notices_helper.rb:
```diff
@@ -0,0 +1,9 @@
+module Me::NoticesHelper
+  def dissmiss_all_notices_button
+    link_to_remote_with_confirm(:dismiss_all_notices.t,
```
I try to avoid `link_to_remote...` in new code and prefer `link_to, ... , remote: true`. This will make the migration away from prototype easier at some point. I will mark the remote functions as deprecated.

aliaksandrb replied:
Okey, need to fix it..
I just wanted neat confirmation modal rather than raw alert-styled by browser.

azul replied:
hmm then we should make it so the confirm option of link_to also uses the nice modal version.

aliaksandrb replied:
But modal is made by another Prototype plugin, isn't it?
If we migrate into jQuery in future we may want to use Bootstrap's modals?
On Jul 11, 2015 5:01 PM, "azul" <notifications@github.com> wrote:

> In app/helpers/me/notices_helper.rb
> <https://github.com/riseuplabs/crabgrass-core/pull/269#discussion_r34412877>
> :
>
> > @@ -0,0 +1,9 @@
> > +module Me::NoticesHelper
> > +  def dissmiss_all_notices_button
> > +    link_to_remote_with_confirm(:dismiss_all_notices.t,
>
> hmm then we should make it so the confirm option of link_to also uses the
> nice modal version.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/riseuplabs/crabgrass-core/pull/269/files#r34412877>.
>

azul replied:
yes. definitly. Being able to use some of the bootstrap js is part of hte motivation for moving to jquery.

elijh replied:
well, azul and I differ here. I like to use link_to_remote. I find it much more sensible than the new form. When we switch to jquery, we can just write our own link_to_remote function that calls link_to(..., :remote => true)

azul commented:
Hope you don't mind the nitty gritty complains about minor things. Just trying to share my thoughts. This pull request made my day :sunny:

aliaksandrb commented:
Haha, of course not! :)
This change appears on user main page, so definitely subject to discussion.

azul commented on app/helpers/me/notices_helper.rb:
```diff
@@ -0,0 +1,9 @@
+module Me::NoticesHelper
+  def dissmiss_all_notices_button
```
Another minor issue.. i think it's 'dismiss' and not 'dissmiss'

elijh commented on config/locales/en/generic.yml:
```diff
@@ -124,6 +124,9 @@ en:
   dismiss: "Dismiss"
+  dismiss_all_notices: "Dismiss All"
+  dismiss_all_notices_confirmation: "Are you sure want to dismiss all notices?
+    <br/>This action cannot be undone."
```
In html5, the correct markup is `<br>`, not `<br/>`. Strange, but true. In general, it is much better to avoid html markup in translations. It just confuses translators and introduces potential security problems. Sadly, it is still used in various places.

In this specific case, I think maybe we should create a separate key for "This action cannot be undone."

aliaksandrb commented:
I have updated the patch:
- typo fixes
- up top pagination block was removed, view simplified and refactored a bit
- removed `<br> ` tag from translations (moved to helper), split phrases into separate ones
- appropriate styles was added

I believe that the rest related to `link_to_remote_with_confirm` usage. How much we against it?
If it deprecated then we need implement remote link with modal confirmation for it, but because `Prototype` itself would be migrated into `jQuery` there is no sense to implement confirmation modal based on`Prototype`.
Could we leave it for now or should we implement it using simple `link_to ... remote: true` with alert-style confirmation?

Personally I am for "leave it as it is" choice until a new ui-helpers based on `jQuery` and `Bootstrap` would not take place.

P.S. Failed Travis seems not to be related with these changes.

elijh commented:
> Personally I am for "leave it as it is" choice until a new ui-helpers based on jQuery and Bootstrap would not take place.

+1

azul commented:
merged. Thanks a lot @aliaksandrb!
